### PR TITLE
Fix layout preferences on Firefox

### LINKED ISSUE #992: Preferences not restored on Firefox 67b4+
To reproduce:

1. Start Firefox 67 beta 4 or later with a clean profile.
2. Navigate to <https://devdocs.io/>.
3. Open Preferences by clicking the link from the welcome page.
4. Check the `Enable dark theme` check box. (Observed: box is checked, dark theme is applied.)
5. Reload the page. (Observed: box is still checked but light theme is applied. Expected: box is checked, dark theme is applied.)
6. Uncheck the `Enable dark theme` check box. (Observed: box is cleared, dark theme is applied. Expected: box is cleared, light theme is applied.)

Sidebar auto-hide is also affected. Did not test every preference setting.

Additional info: mozregression points to [this set of commits][1] in Firefox that has something to deal with cookies.

[1]: https://hg.mozilla.org/integration/autoland/pushloghtml?fromchange=615a12a9276839868bef0858f3fb535a04106295&tochange=3cddc7cd4da5fcd26b93becb55a028621ab68f64

### PR DESCRIPTION
**Pull request by jmerle**
Currently layout preferences are loaded based on the cookies sent in the HTTP request to devdocs.io. This does not work on recent versions of Firefox, as has been reported in #992.

After this PR is merged, layout preferences are also loaded locally based on the user's stored cookies. By loading preferences client-side, we no longer need to rely on the optional [`Cookie`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cookie) header. This fixes #992.

Locally loading layout preferences happens before the content is shown, so there is no flickering between light and dark themes on slow connections. Performance does not seem to be affected neither.

The code to load the layout preferences based on the cookies sent with the HTTP request has not been removed as there seem to be some dependencies between that code and the AppCache-related code.

### CODE REVIEW

**Comment by sbrl** [comment]
Hey, awesome! Is there a timeline for when this might get merged? Looking at the list of other PRs though it seems highly unlikely this will actually get merged in the next few months though :-/

**Comment by lostfictions** [comment]
Would love to see this merged! It's getting a bit tedious to toggle and retoggle dark mode every time I open DevDocs in Firefox (ie. fifty times a day :) )

**Comment by jmerle** [comment]
> Hey, awesome! Is there a timeline for when this might get merged? Looking at the list of other PRs though it seems highly unlikely this will actually get merged in the next few months though :-/

There is no exact timeline, but I'm pretty sure we can get it merged & deployed in about 2-4 weeks, which is already a lot better than those few months :smiley:. We've had a maintainers chat about a month ago in which we discussed the future of DevDocs (now that the original creator is no longer working on it) and we came up with a plan for the next deployment. Basically, we want to get all non-scraper PR's (except for #986) merged first, after which we can get the scraper PR's going. This comes down to this PR and all those with either the [feature](https://github.com/freeCodeCamp/devdocs/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+label%3Afeature) or the [improvement](https://github.com/freeCodeCamp/devdocs/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+label%3Aimprovement) label. I have already reviewed all PR's that aren't mine, and @j-f1 can probably review my PR's next week.

After these PR's are reviewed we can (finally!) run our first deployment since the project got handed over to FreeCodeCamp, with hopefully many more to come.

**Comment by sbrl** [comment]
Thanks so much for the update, @jmerle!

I can tell it's been a long road - I'm excited for that update!

File: assets/javascripts/views/content/settings_page.coffee [hunk 1]
```diff
@@ -17,23 +14,18 @@ class app.views.SettingsPage extends app.View
     settings.dark = app.settings.get('dark')
     settings.smoothScroll = !app.settings.get('fastScroll')
     settings.arrowScroll = app.settings.get('arrowScroll')
-    settings[layout] = app.settings.hasLayout(layout) for layout in LAYOUTS
+    settings[layout] = app.settings.hasLayout(layout) for layout in app.settings.getAllLayouts()
```
Comment (j-f1) [inline assets/javascripts/views/content/settings_page.coffee]: Why not refer to the `LAYOUTS` variable directly?

```suggestion
    settings[layout] = app.settings.hasLayout(layout) for layout in app.settings.LAYOUTS
```

This requires a small change to the definition:

```coffee
LAYOUTS: ['_max-width', '_sidebar-hidden', '_native-scrollbars']
```

and the single use in the `Settings` class:

```coffee
@toggleLayout(layout, @hasLayout(layout)) for layout in @LAYOUTS

Comment (jmerle) [reply assets/javascripts/views/content/settings_page.coffee]: Thanks, fixed!

**Comment by jmerle** [comment]
Superseded by #1022.

**Pull request closed without merging**